tests/turnstile_complex.c

```diff
@@ -0,0 +1,171 @@
+#include <ktest.h>
+#include <mutex.h>
+#include <runq.h>
+#include <sched.h>
+#include <thread.h>
+
+#define T 4
+
+#define assert_priorities(p0, p1, p2, p3)                                      \
+  assert(T >= 4 && td[0]->td_prio == p0 && td[1]->td_prio == p1 &&             \
+         td[2]->td_prio == p2 && td[3]->td_prio == p3)
+
+#define lend_prio(td, prio)                                                    \
+  do {                                                                         \
+    WITH_SPINLOCK(td->td_spin) {                                               \
+      sched_lend_prio(td, prio);                                               \
+    }                                                                          \
+  } while (0)
+
+#define unlend_prio(td, prio)                                                  \
+  do {                                                                         \
+    WITH_SPINLOCK(td->td_spin) {                                               \
+      sched_unlend_prio(td, prio);                                             \
+    }                                                                          \
+  } while (0)
+
+#define set_base_prio(td, prio)                                                \
+  do {                                                                         \
+    WITH_SPINLOCK(td->td_spin) {                                               \
+      td->td_base_prio = prio;                                                 \
+    }                                                                          \
+  } while (0)
+
+static mtx_t *mtx1 = &MTX_INITIALIZER(MTX_DEF);
+static mtx_t *mtx2 = &MTX_INITIALIZER(MTX_DEF);
+static thread_t *td[T];
+static volatile bool low_prio_mtx1_acquired;
+static volatile bool med_prio_mtx1_acquired;
+static volatile bool high_prio_mtx2_acquired;
+
+enum {
+  /* Priorities are multiplies of RunQueue_PriorityPerQueue
+   * so that each priority matches different run queue. */
+  LOW = 0,
+  MED = RQ_PPQ,
+  HIGH = 2 * RQ_PPQ
+};
+
+/* td3 */
+static void high_prio_task(void *arg) {
+  WITH_MTX_LOCK (mtx2) {
+    /* When we get here, low_prio_task and med_prio_task1 should've
+       got their mutexes and all priorities should be in initial state. */
+    high_prio_mtx2_acquired = 1;
+    assert(low_prio_mtx1_acquired == 1);
+    assert(med_prio_mtx1_acquired == 1);
+
+    assert_priorities(LOW, MED, MED, HIGH);
+  }
+}
+
+/* td1 */
+static void med_prio_task1(void *arg) {
+  WITH_NO_PREEMPTION {
+    WITH_MTX_LOCK (mtx2) {
+      WITH_MTX_LOCK (mtx1) {
+        /* Td0 (low_prio_task) has released mtx1.
+         * Td3 (high_prio_task) is waiting for mtx2. */
```

Comment:
Thread `td0` has released `mtx1` because we are now in `mtx1`. It doesn't need any assert.
Thread `td3` is waiting for `mtx2` - what about `assert(td[3]->td_blocked->ts_wchan == mtx2);`?

Reply:
However we don't know who owned `mtx1`. Assert about `td0` and `mtx1` could go right before the attempt to lock `mtx1` - something like `assert(mtx1_owner == td0)`. Well, we still wouldn't know if somebody didn't get the mutex after `td0` unlocked it and before `td3` acquired it.
About the second one: looks okay. I'd alternatively consider `assert(td[3]->td_wchan == mtx2)`.

Reply:
1a) Now we do know who owned mtx1. `assert(mtx1->m_owner == td[0]);`
1b) My comment doesn't say that we were the first to get the mutex after `td0` released it.
2) Done.